docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,500 @@
+Tn`, and `C` is a list of subtype and supertype constraints on the `Ti`.
+
+This relation can be read as "`P` is a subtype match for `Q` with respect to the
+list of type variables `L` under constraints `C`".
+
+
+By invariant, at any point in constraint generation, only one of `P` and `Q` may
+be a type schema (that is, contain `?`), only one of `P` and `Q` may contain any
+of the `Ti`, and neither may contain both.  That is, constraint generation is a
+relation on type-schema/type pairs and type/type-schema pairs, only the type
+element of which may refer to the `Ti`.
+
+### Notes:
+
+- For convenience, ordering matters in this presentation: where any two clauses
+  overlap syntactically, the first match is preferred.
+- This presentation is assuming appropriate well-formedness conditions on the
+  input types (e.g. non-cyclic class hierarchies)
+
+### Syntactic notes:
+
+- `C0 + C1` is the concatenation of constraint lists `C0` and `C1`.
+
+### Rules
+
+- The unknown type `?` is a subtype match for any type `Q` with no constraints.
+- Any type `P` is a subtype match for the unknown type `?` with no constraints.
+- A type variable `T` in `L` is a subtype match for any type schema `Q`:
+  - Under constraint `T <: Q`.
+- A type schema `Q` is a subtype match for a type variable `T` in `L`:
+  - Under constraint `Q <: T`.
+- Any two equal types `P` and `Q` are subtype matches under no constraints.
+- Any type `P` is a subtype match for `dynamic`, `Object`, or `void` under no
+  constraints.
+- `Null` is a subtype match for any type `Q` under no constraints.
+- `FutureOr<P>` is a subtype match for `FutureOr<Q>` with respect to `L` under
+  constraints `C`:
+  - If `P` is a subtype match for `Q` with respect to `L` under constraints `C`.
+- `FutureOr<P>` is a subtype match for `Q` with respect to `L` under
+constraints `C0 + C1`.
+  - If `Future<P>` is a subtype match for `Q` with respect to `L` under
+    constraints `C0`.
+  - And `P` is a subtype match for `Q` with respect to `L` under constraints
+    `C1`.
+- `P` is a subtype match for `FutureOr<Q>` with respect to `L` under constraints
+  `C`:
+  - If `P` is a subtype match for `Future<Q>` with respect to `L` under
+    constraints `C`.
+  - Or `P` is not a subtype match for `Future<Q>` with respect to `L` under
+    constraints `C`
+    - And `P` is a subtype match for `Q` with respect to `L` under constraints
+      `C`
+- A type variable `T` not in `L` with bound `P` is a subtype match for the same
+type variable `T` with bound `Q` with respect to `L` under constraints `C`:
+  - If `P` is a subtype match for `Q` with respect to `L` under constraints `C`.
+- A type variable `T` not in `L` with bound `P` is a subtype match for a type
+`Q` with respect to `L` under constraints `C`:
+  - If `P` is a subtype match for `Q` with respect to `L` under constraints `C`.
+- A type `P<M0, ..., Mk>` is a subtype match for `P<N0, ..., Nk>` with respect
```

Comment:
Are we missing a case here?  Something like:

A type `P` is a subtype match for a type varible `T` with respect to `L` under constraints `C`: ...

Or maybe this case is *never* a subtype match?